README.md

# koa-logger

[![npm version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]

 Development style logger middleware for [koa](https://github.com/koajs/koa).  Compatible with [request-received](https://github.com/cabinjs/request-received).

___Notice: `koa-logger@2` supports `koa@2`; if you want to use this module with `koa@1`, please use `koa-logger@1`.___

```
<-- GET /
--> GET / 200 835ms 746b
<-- GET /
--> GET / 200 960ms 1.9kb
<-- GET /users
--> GET /users 200 357ms 922b
<-- GET /users?page=2
--> GET /users?page=2 200 466ms 4.66kb
```

## Installation

```js
$ npm install koa-logger
```

## Example

```js
const logger = require('koa-logger')
const Koa = require('koa')

const app = new Koa()
app.use(logger())
```

## Notes

  Recommended that you `.use()` this middleware near the top
  to "wrap" all subsequent middleware.

## Use Custom Transporter

```js
const logger = require('koa-logger')
const Koa = require('koa')

const app = new Koa()
app.use(logger((str, args) => {
  // redirect koa logger to other output pipe
  // default is process.stdout(by console.log function)
}))
```
or
```js
app.use(logger({
  transporter: (str, args) => {
    // ...
  }
}))
```

  Param `str` is output string with ANSI Color, and you can get pure text with other modules like `strip-ansi`  
  Param `args` is a array by `[format, method, url, status, time, length]`

## License

  MIT

[npm-image]: https://img.shields.io/npm/v/koa-logger.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/koa-logger
[travis-image]: https://img.shields.io/travis/koajs/logger.svg?style=flat-square
[travis-url]: https://travis-ci.org/koajs/logger


History.md

3.1.0 / 2017-09-26
==================

 * Adds Boom compatibility (#64)

3.0.1 / 2017-06-30
==================

 * Fixes error 'chalk[color] does not return function' when using custom (#61)

3.0.0 / 2017-05-16
==================

 * use async function
 * update bytes to 2.x

2.0.1 / 2017-01-12
==================

 * add color for code 0
 * bump deps

2.0.0 / 2015-11-19
==================

* add: support for koa 2
* remove: support for koa 1

1.2.2 / 2014-07-05
==================

 * fix: stop using octal literals for strict mode

1.2.1 / 2014-05-13
==================

 * bump bytes

1.2.0 / 2014-02-12
==================

 * add: add colors and symbols for error and close events
 * remove: only make the method bold

1.1.1 / 2014-01-15
==================

 * fix: use `ctx.originalUrl` instead of `ctx.url`

1.1.0 / 2013-12-21
==================

 * add: log the response length of streams
 * add: humanize the response time
 * add: log when response is finished/closed, not caught upstream

1.0.1 / 2013-08-22
==================

 * fix exception handling


test.js
'use strict'
/**
 * test cases
 */

// test tools
const chai = require('chai')
const sinon = require('sinon')
const sc = require('sinon-chai')
const request = require('supertest')
chai.use(sc)
const expect = chai.expect

// test subjects
const chalk = require('chalk')
const app = require('./test-server')
let log, sandbox

describe('koa-logger', function () {
  beforeEach(function () {
    sandbox = sinon.sandbox.create()
    log = sandbox.spy(console, 'log')
  })

  afterEach(function () {
    sandbox.restore()
  })

  it('should log a request', function (done) {
    request(app.listen()).get('/200').expect(200, 'hello world', function () {
      expect(log).to.have.been.called // eslint-disable-line
      done()
    })
  })

  it('should log a request with correct method and url', function (done) {
    request(app.listen()).head('/200').expect(200, function () {
      expect(log).to.have.been.calledWith('  ' + chalk.gray('<--') +
        ' ' + chalk.bold('%s') +
        ' ' + chalk.gray('%s'),
          'HEAD',
          '/200')
      done()
    })
  })

  it('should log a response', function (done) {
    request(app.listen()).get('/200').expect(200, function () {
      expect(log).to.have.been.calledTwice // eslint-disable-line
      done()
    })
  })

  it('should log a 200 response', function (done) {
    request(app.listen()).get('/200').expect(200, function () {
      expect(log).to.have.been.calledWith('  ' + chalk.gray('-->') +
        ' ' + chalk.bold('%s') +
        ' ' + chalk.gray('%s') +
        ' ' + chalk.green('%s') +
        ' ' + chalk.gray('%s') +
        ' ' + chalk.gray('%s'),
          'GET',
          '/200',
          200,
          sinon.match.any,
          '11b')
      done()
    })
  })

  it('should log a 301 response', function (done) {
    request(app.listen()).get('/301').expect(301, function () {
      expect(log).to.have.been.calledWith('  ' + chalk.gray('-->') +
        ' ' + chalk.bold('%s') +
        ' ' + chalk.gray('%s') +
        ' ' + chalk.cyan('%s') +
        ' ' + chalk.gray('%s') +
        ' ' + chalk.gray('%s'),
          'GET',
          '/301',
          301,
          sinon.match.any,
          '-')
      done()
    })
  })

  it('should log a 304 response', function (done) {
    request(app.listen()).get('/304').expect(304, function () {
      expect(log).to.have.been.calledWith('  ' + chalk.gray('-->') +
        ' ' + chalk.bold('%s') +
        ' ' + chalk.gray('%s') +
        ' ' + chalk.cyan('%s') +
        ' ' + chalk.gray('%s') +
        ' ' + chalk.gray('%s'),
          'GET',
          '/304',
          304,
          sinon.match.any,
          '')
      done()
    })
  })

  it('should log a 404 response', function (done) {
    request(app.listen()).get('/404').expect(404, function () {
      expect(log).to.have.been.calledWith('  ' + chalk.gray('-->') +
        ' ' + chalk.bold('%s') +
        ' ' + chalk.gray('%s') +
        ' ' + chalk.yellow('%s') +
        ' ' + chalk.gray('%s') +
        ' ' + chalk.gray('%s'),
          'GET',
          '/404',
          404,
          sinon.match.any,
          '9b')
      done()
    })
  })

  it('should log a 500 response', function (done) {
    request(app.listen()).get('/500').expect(500, function () {
      expect(log).to.have.been.calledWith('  ' + chalk.gray('-->') +
        ' ' + chalk.bold('%s') +
        ' ' + chalk.gray('%s') +
        ' ' + chalk.red('%s') +
        ' ' + chalk.gray('%s') +
        ' ' + chalk.gray('%s'),
          'GET',
          '/500',
          500,
          sinon.match.any,
          '12b')
      done()
    })
  })

  it('should log middleware error', function (done) {
    request(app.listen()).get('/error').expect(500, function () {
      expect(log).to.have.been.calledWith('  ' + chalk.red('xxx') +
        ' ' + chalk.bold('%s') +
        ' ' + chalk.gray('%s') +
        ' ' + chalk.red('%s') +
        ' ' + chalk.gray('%s') +
        ' ' + chalk.gray('%s'),
          'GET',
          '/error',
          500,
          sinon.match.any,
          '-')
      done()
    })
  })

  it('should log a 500 response with boom', function (done) {
    request(app.listen()).get('/500-boom').expect(500, function () {
      expect(log).to.have.been.calledWith('  ' + chalk.red('xxx') +
        ' ' + chalk.bold('%s') +
        ' ' + chalk.gray('%s') +
        ' ' + chalk.red('%s') +
        ' ' + chalk.gray('%s') +
        ' ' + chalk.gray('%s'),
          'GET',
          '/500-boom',
          500,
          sinon.match.any,
          '-')
      done()
    })
  })
})


example.js

const logger = require('./')
const Koa = require('koa')
const compress = require('koa-compress')()
const app = new Koa()

// wrap subsequent middleware in a logger

app.use(logger())

// 204

app.use(function (ctx, next) {
  if (ctx.path === '/204') ctx.status = 204
  else return next()
})

// 404

app.use(function (ctx, next) {
  if (ctx.path === '/404') return
  return next()
})

// destroy

app.use(function (ctx, next) {
  if (ctx.path === '/close') return ctx.req.destroy()
  return next()
})

// compress the response 1/2 the time to calculate the stream length

app.use(function (ctx, next) {
  if (Math.random() > 0.5) {
    return next()
  } else {
    return compress(ctx, next)
  }
})

// response middleware

app.use(function (ctx, next) {
  // yield control downstream
  next().then(function () {
    // sleep for 0-2s
    return sleep(Math.random() * 2000 | 0)
  }).then(function () {
    // error
    if (Math.random() > 0.75) {
      const err = new Error('boom')
      err.status = 500
      throw err
    }

    // random body
    const body = Array(Math.random() * 5 * 1024 | 9).join('a')
    ctx.status = 200
    ctx.body = body
  })
})

const port = process.env.PORT || 3000
app.listen(port)
console.log('listening on port ' + port)

// sleep helper

function sleep (ms) {
  return new Promise(function (resolve) {
    setTimeout(resolve, ms)
  })
}


index.js
/**
 * Module dependencies.
 */
'use strict'

const Counter = require('passthrough-counter')
const humanize = require('humanize-number')
const bytes = require('bytes')
const chalk = require('chalk')

/**
 * Expose logger.
 */

module.exports = dev

/**
 * Color map.
 */

const colorCodes = {
  7: 'magenta',
  5: 'red',
  4: 'yellow',
  3: 'cyan',
  2: 'green',
  1: 'green',
  0: 'yellow'
}

/**
 * Development logger.
 */

function dev (opts) {
  return async function logger (ctx, next) {
    // request
    const start = Date.now()
    console.log('  ' + chalk.gray('<--') +
      ' ' + chalk.bold('%s') +
      ' ' + chalk.gray('%s'),
        ctx.method,
        ctx.originalUrl)

    try {
      await next()
    } catch (err) {
      // log uncaught downstream errors
      log(ctx, start, null, err)
      throw err
    }

    // calculate the length of a streaming response
    // by intercepting the stream with a counter.
    // only necessary if a content-length header is currently not set.
    const length = ctx.response.length
    const body = ctx.body
    let counter
    if (length == null && body && body.readable) {
      ctx.body = body
        .pipe(counter = Counter())
        .on('error', ctx.onerror)
    }

    // log when the response is finished or closed,
    // whichever happens first.
    const res = ctx.res

    const onfinish = done.bind(null, 'finish')
    const onclose = done.bind(null, 'close')

    res.once('finish', onfinish)
    res.once('close', onclose)

    function done (event) {
      res.removeListener('finish', onfinish)
      res.removeListener('close', onclose)
      log(ctx, start, counter ? counter.length : length, null, event)
    }
  }
}

/**
 * Log helper.
 */

function log (ctx, start, len, err, event) {
  // get the status code of the response
  const status = err
    ? (err.isBoom ? err.output.statusCode : err.status || 500)
    : (ctx.status || 404)

  // set the color of the status code;
  const s = status / 100 | 0
  const color = colorCodes.hasOwnProperty(s) ? colorCodes[s] : 0

  // get the human readable response length
  let length
  if (~[204, 205, 304].indexOf(status)) {
    length = ''
  } else if (len == null) {
    length = '-'
  } else {
    length = bytes(len).toLowerCase()
  }

  const upstream = err ? chalk.red('xxx')
    : event === 'close' ? chalk.yellow('-x-')
    : chalk.gray('-->')

  console.log('  ' + upstream +
    ' ' + chalk.bold('%s') +
    ' ' + chalk.gray('%s') +
    ' ' + chalk[color]('%s') +
    ' ' + chalk.gray('%s') +
    ' ' + chalk.gray('%s'),
      ctx.method,
      ctx.originalUrl,
      status,
      time(start),
      length)
}

/**
 * Show the response time in a human readable format.
 * In milliseconds if less than 10 seconds,
 * in seconds otherwise.
 */

function time (start) {
  const delta = Date.now() - start
  return humanize(delta < 10000
    ? delta + 'ms'
    : Math.round(delta / 1000) + 's')
}
